lib/ipip.c

```diff
@@ -131,3 +131,67 @@ encapsulate(struct rte_mbuf *pkt, uint8_t priority,
 
 	return 0;
 }
+
+int
+decapsulate(struct rte_mbuf *pkt, uint8_t *priority,
+	struct ipip_tunnel_info *info)
+{
+	uint16_t ethertype;
+	uint8_t l4_proto;
+	uint16_t outer_header_len = 0;
+	struct ether_hdr *eth_hdr = rte_pktmbuf_mtod(pkt, struct ether_hdr *);
+	struct ipv4_hdr *ipv4_hdr;
+	struct ipv6_hdr *ipv6_hdr;
+
+	outer_header_len += sizeof(struct ether_hdr);
+	ethertype = rte_be_to_cpu_16(eth_hdr->ether_type);
+
+	switch (ethertype) {
+	case ETHER_TYPE_IPv4:
+		ipv4_hdr = (struct ipv4_hdr *)
+			((char *)eth_hdr + outer_header_len);
+		outer_header_len += sizeof(struct ipv4_hdr);
+		*priority = (ipv4_hdr->type_of_service >> 2);
+		l4_proto = ipv4_hdr->next_proto_id;
+		if (*priority >= 2) {
+			info->flow.proto = ETHER_TYPE_IPv4;
+			info->flow.f.v4.src = ipv4_hdr->src_addr;
+			info->flow.f.v4.dst = ipv4_hdr->dst_addr;
+		}
+		break;
+	case ETHER_TYPE_IPv6:
+		ipv6_hdr = (struct ipv6_hdr *)
```

Comment:
This cast is done without knowing that the packet is long enough to hold an IPv6 header.